tokio/src/sync/once_cell.rs

```diff
@@ -304,32 +315,135 @@ impl<T> OnceCell<T> {
+        }
+    }
+
+    /// Notify all waiting tasks.
+    fn notify_initialized(&self) {
+        let mut wakers = WakeList::new();
+        let mut waiters = self.waiters.lock();
+        loop {
+            while wakers.can_push() {
+                if let Some(mut waiter) = waiters.pop_front() {
+                    // Safety: we hold the lock, so we can access the waker.
+                    if let Some(waker) = unsafe { waiter.as_mut().waker.take() } {
+                        wakers.push(waker);
+                    }
+                } else {
+                    // Release the lock before notifying
+                    drop(waiters);
+                    wakers.wake_all();
+                    return;
+                }
             }
-            Err(TryAcquireError::Closed) => {
-                // The semaphore was closed. Some other task has initialized
-                // the value.
-                Err(SetError::AlreadyInitializedError(value))
+            drop(waiters);
+            wakers.wake_all();
+            waiters = self.waiters.lock();
+        }
+    }
+
+    /// Notify one initializer task.
+    fn notify_unlocked(&self) {
+        let mut waiters = self.waiters.lock();
+        if let Some(mut waiter) = waiters.drain_filter(|w| w.is_initializer).next() {
+            // SAFETY: we hold the lock, so we can access the waker.
+            let waker = unsafe { waiter.as_mut().waker.take() };
+            drop(waiters);
+            waker.unwrap().wake();
+        }
+    }
+
+    /// Wait for the state to have the required value.
+    /// If the node is pushed in the waiter
+    fn poll_wait(&self, cx: &mut Context<'_>, node: Pin<&mut Waiter>) -> Poll<u8> {
+        // Clone the waker before locking, a waker clone can be triggering arbitrary code.
+        let waker = cx.waker().clone();
+        let mut waiters = self.waiters.lock();
+        // SAFETY: node is not moved out of pinned reference,
+        // and it is accessed with the lock held
+        let node = unsafe { node.get_unchecked_mut() };
+        // Checks the state with the lock held; if the waiter has been notified,
+        // then the state is ensured to have been modified.
+        // Uses acquire ordering as value may be accessed after in case of success.
+        let state = self.state.load(Ordering::Acquire);
+        if state == STATE_SET || node.is_initializer && state == STATE_UNSET {
+            // If there is a waker, it means the node is queued,
+            // so it is removed from the list.
+            if node.waker.take().is_some() {
+                // SAFETY: The node is contained in the list.
+                unsafe { waiters.remove(NonNull::from(node)) };
+            }
+            Poll::Ready(state)
+        } else {
+            // If there is no waker, it means the node is not queued,
+            // so it is pushed in the list.
+            if node.waker.replace(waker).is_none() {
```

Comment:
Here (and above) you run the destructor of the waker with the lock held.

Reply:
Fixed in https://github.com/tokio-rs/tokio/pull/7161/commits/cac4ee0ca2354d2fa30c041ea00dca72da80418f